File: coinswap-kotlin/lib/src/test/kotlin/org/coinswap/SwapTest.kt
/**
 * JVM integration test: 4 takers × 2 makers.
 *
 * Mirrors the Rust `swap_test`: one test per (backend × protocol) combination —
 * legacy/taproot over rpc/electrum — each running a 2-maker coinswap against the
 * Docker regtest stack (1 RPC maker + 1 Electrum maker).
 */

package org.coinswap

import org.junit.jupiter.api.Test
import org.junit.jupiter.api.io.TempDir
import java.nio.file.Path
import kotlin.test.assertEquals
import kotlin.test.assertNotNull
import kotlin.test.assertTrue

class SwapTest {

    private enum class Backend { RPC, ELECTRUM }

    private val swapAmount = 500_000uL

    /** Fund [address] with [btc] BTC from the Docker bitcoind `test` wallet. */
    private fun fund(address: String, btc: String) {
        val p = ProcessBuilder(
            "docker", "exec", "coinswap-bitcoind",
            "bitcoin-cli", "-regtest", "-rpcport=18442",
            "-rpcwallet=test", "-rpcuser=user", "-rpcpassword=password",
            "sendtoaddress", address, btc,
        ).redirectErrorStream(true).start()
        val out = p.inputStream.bufferedReader().readText().trim()
        check(p.waitFor() == 0) { "funding failed: $out" }
    }

    /** Sync until spendable reaches [target], tolerating Electrum indexing lag. */
    private fun waitForSpendable(taker: Taker, target: ULong): Balances {
        repeat(30) {
            taker.syncAndSave()
            val b = taker.getBalances()
            if (b.spendable.toULong() >= target) return b
            Thread.sleep(3000)
        }
        return taker.getBalances()
    }

    /** Run one taker end-to-end: init → fund → sync → 2-maker coinswap → assert. */
    private fun runSwap(
        name: String,
        dataDir: Path,
        backend: Backend,
        protocol: String,
        addrType: String,
    ) {
        println("\n=== $name ($protocol) ===")

        val rpcConfig = if (backend == Backend.RPC) {
            RpcConfig("localhost:18442", "user", "password", "kotlin_$name")
        } else null
        val backendConfig = if (backend == Backend.ELECTRUM) {
            BackendConfig(kind = "electrum", url = "tcp://localhost:50001")
        } else null

        val taker = Taker.init(
            dataDir = dataDir.toString(),
            walletFileName = name,
            rpcConfig = rpcConfig,
            controlPort = 9051u,
            torAuthPassword = "coinswap",
            zmqAddr = "tcp://localhost:28332",
            password = "",
            nostrRelays = null,
            backendConfig = backendConfig,
        )

        taker.syncOfferbookAndWait()

        // Fund with 2x the swap amount across 4 fresh addresses.
        val quarterBtc = "0.0025" // 250,000 sats; 4x = 1,000,000 = 2 * swapAmount
        repeat(4) {
            val addr = taker.getNextExternalAddress(AddressType(addrType)).addr
            fund(addr, quarterBtc)
        }
        val funded = waitForSpendable(taker, swapAmount * 2uL)
        assertEquals(
            (swapAmount * 2uL).toLong(), funded.spendable,
            "$name: spendable should equal funded amount",
        )

        val swapId = taker.prepareCoinswap(
            SwapParams(
                protocol = protocol,
                sendAmount = swapAmount,
                makerCount = 2u,
                txCount = 1u,
                requiredConfirms = 1u,
                manuallySelectedOutpoints = null,
                preferredMakers = null,
            ),
        )
        val report = taker.startCoinswap(swapId)
        assertNotNull(report)
        assertEquals(2u, report.makersCount, "$name: should route through 2 makers")
        assertTrue(
            report.status.uppercase().contains("SUCCESS"),
            "$name: swap status was ${report.status}",
        )
        println("✓ $name passed (swap_id ${report.swapId})")
    }

    @Test
    fun `legacy rpc swap`(@TempDir dir: Path) =
        runSwap("legacy_rpc", dir, Backend.RPC, "Legacy", "P2WPKH")

    @Test
    fun `taproot rpc swap`(@TempDir dir: Path) =
        runSwap("taproot_rpc", dir, Backend.RPC, "Taproot", "P2TR")

    @Test
    fun `legacy electrum swap`(@TempDir dir: Path) =
        runSwap("legacy_electrum", dir, Backend.ELECTRUM, "Legacy", "P2WPKH")

    @Test
    fun `taproot electrum swap`(@TempDir dir: Path) =
        runSwap("taproot_electrum", dir, Backend.ELECTRUM, "Taproot", "P2TR")
}

File: coinswap-csharp/Coinswap.Tests/SwapTest.cs
using System.Diagnostics;
using Coinswap.Native;
using Xunit;
using Xunit.Abstractions;

namespace Coinswap.Tests;

/// <summary>
/// Live integration test: 4 takers × 2 makers.
///
/// One test per (backend × protocol) combination — legacy/taproot over
/// rpc/electrum — each running a 2-maker coinswap against the Docker regtest
/// stack (1 RPC maker + 1 Electrum maker). Mirrors the Rust <c>swap_test</c> and
/// the Kotlin <c>SwapTest</c>. Requires the stack running:
/// <c>cd ../ffi-commons &amp;&amp; ./ffi-docker-setup start</c>.
/// </summary>
public class SwapTest
{
    private const string BitcoindContainer = "coinswap-bitcoind";
    private const string RpcUrl = "localhost:18442";
    private const string RpcUser = "user";
    private const string RpcPassword = "password";
    private const string ZmqAddr = "tcp://localhost:28332";
    private const string ElectrumUrl = "tcp://localhost:50001";
    private const ushort ControlPort = 9051;

    /// <summary>Amount swapped by each taker, in sats. The taker is funded with 2×.</summary>
    private const ulong SwapAmount = 500_000;

    private enum Backend { Rpc, Electrum }

    private readonly ITestOutputHelper _out;

    public SwapTest(ITestOutputHelper output) => _out = output;

    [Fact]
    public void LegacyRpc() =>
        RunSwap("legacy_rpc", Backend.Rpc, "Legacy", "P2WPKH");

    [Fact]
    public void TaprootRpc() =>
        RunSwap("taproot_rpc", Backend.Rpc, "Taproot", "P2TR");

    [Fact]
    public void LegacyElectrum() =>
        RunSwap("legacy_electrum", Backend.Electrum, "Legacy", "P2WPKH");

    [Fact]
    public void TaprootElectrum() =>
        RunSwap("taproot_electrum", Backend.Electrum, "Taproot", "P2TR");

    /// <summary>Run one taker end-to-end: init → fund → sync → 2-maker coinswap → assert.</summary>
    private void RunSwap(string name, Backend backend, string protocol, string addrType)
    {
        _out.WriteLine($"\n=== {name} ({protocol}) ===");

        var dataDir = Path.Combine(Path.GetTempPath(), $"coinswap-csharp-{name}-{Guid.NewGuid():N}");
        Directory.CreateDirectory(dataDir);

        var rpcConfig = backend == Backend.Rpc
            ? new RpcConfig(
                Url: RpcUrl,
                Username: RpcUser,
                Password: RpcPassword,
                WalletName: $"csharp_{name}")
            : null;

        var backendConfig = backend == Backend.Electrum
            ? new BackendConfig(Kind: "electrum", Url: ElectrumUrl)
            : null;

        // Positional args mirror the Rust `Taker::init` signature order:
        // (data_dir, wallet_file_name, rpc_config, control_port, tor_auth_password,
        //  zmq_addr, password, nostr_relays, backend_config).
        using var taker = Taker.Init(
            dataDir,
            name,
            rpcConfig,
            ControlPort,
            "coinswap",
            ZmqAddr,
            "",
            null,
            backendConfig);

        taker.SyncOfferbookAndWait();

        // Fund with 2× the swap amount across 4 fresh external addresses.
        const string quarterBtc = "0.0025"; // 250,000 sats; 4× = 1,000,000 = 2 × SwapAmount
        for (var i = 0; i < 4; i++)
        {
            var addr = taker.GetNextExternalAddress(new AddressType(addrType)).Addr;
            Fund(addr, quarterBtc);
        }

        var funded = WaitForSpendable(taker, SwapAmount * 2);
        Assert.True(
            (ulong)funded.Spendable >= SwapAmount * 2,
            $"{name}: spendable ({funded.Spendable}) should reach funded amount ({SwapAmount * 2})");

        var swapId = taker.PrepareCoinswap(new SwapParams(
            Protocol: protocol,
            SendAmount: SwapAmount,
            MakerCount: 2,
            TxCount: 1,
            RequiredConfirms: 1,
            ManuallySelectedOutpoints: null,
            PreferredMakers: null));

        var report = taker.StartCoinswap(swapId);
        Assert.NotNull(report);
        Assert.Equal(2u, report.MakersCount);
        Assert.Contains("SUCCESS", report.Status.ToUpperInvariant());

        _out.WriteLine($"✓ {name} passed (swap_id {report.SwapId})");
    }

    /// <summary>Sync until spendable reaches <paramref name="target"/>, tolerating Electrum indexing lag.</summary>
    private Balances WaitForSpendable(Taker taker, ulong target)
    {
        for (var i = 0; i < 30; i++)
        {
            taker.SyncAndSave();
            var b = taker.GetBalances();
            if ((ulong)b.Spendable >= target) return b;
            Thread.Sleep(3000);
        }
        return taker.GetBalances();
    }

    /// <summary>Sends BTC to the taker address from the docker-hosted "test" wallet.</summary>
    private void Fund(string address, string amountBtc)
    {
        var (code, _, stderr) = RunDocker(
            "exec", BitcoindContainer, "bitcoin-cli", "-regtest", "-rpcport=18442",
            "-rpcwallet=test", $"-rpcuser={RpcUser}", $"-rpcpassword={RpcPassword}",
            "sendtoaddress", address, amountBtc);
        if (code != 0)
            throw new InvalidOperationException($"funding failed: {stderr}");
    }

    private static (int code, string stdout, string stderr) RunDocker(params string[] args)
    {
        var psi = new ProcessStartInfo("docker")
        {
            RedirectStandardOutput = true,
            RedirectStandardError = true,
            UseShellExecute = false,
        };
        foreach (var a in args) psi.ArgumentList.Add(a);

        using var p = Process.Start(psi)!;
        var stdout = p.StandardOutput.ReadToEnd();
        var stderr = p.StandardError.ReadToEnd();
        p.WaitForExit();
        return (p.ExitCode, stdout, stderr);
    }
}
